build.sh

```diff
@@ -1,96 +1,353 @@
+    __buildlog=$__scriptpath/msbuild.log
+
+    MONO29679=1 ReferenceAssemblyRoot=$__referenceassemblyroot mono $__msbuildpath "$__buildproj" /nologo /verbosity:minimal "/fileloggerparameters:Verbosity=normal;LogFile=$__buildlog" /t:Build /p:OSGroup=$__BuildOS /p:UseRoslynCompiler=true /p:COMPUTERNAME=$(hostname) /p:USERNAME=$(id -un) "$@"
+    BUILDERRORLEVEL=$?
+
+    echo
+
+    # Pull the build summary from the log file
+    tail -n 4 "$__buildlog"
+    echo Build Exit Code = $BUILDERRORLEVEL
+}
+
+build_native_corefx()
+{
+    # All set to commence the build
+
+    echo "Commencing build of corefx native components for $__BuildOS.$__BuildArch.$__BuildType"
+    cd "$__IntermediatesDir"
+
+    # Regenerate the CMake solution
+    echo "Invoking cmake with arguments: \"$__nativeroot\" $__CMakeArgs"
+    "$__nativeroot/gen-buildsys-clang.sh" "$__nativeroot" $__ClangMajorVersion $__ClangMinorVersion $__CMakeArgs
+
+    # Check that the makefiles were created.
+
+    if [ ! -f "$__IntermediatesDir/Makefile" ]; then
+        echo "Failed to generate native component build project!"
+        exit 1
+    fi
+
+    # Get the number of processors available to the scheduler
+    # Other techniques such as `nproc` only get the number of
+    # processors available to a single process.
+    if [ `uname` = "FreeBSD" ]; then
+        NumProc=`sysctl hw.ncpu | awk '{ print $2+1 }'`
+    else
+        NumProc=$(($(getconf _NPROCESSORS_ONLN)+1))
+    fi
+
+    # Build
+
+    echo "Executing make install -j $NumProc $__UnprocessedBuildArgs"
+
+    make install -j $NumProc $__UnprocessedBuildArgs
+    if [ $? != 0 ]; then
+        echo "Failed to build corefx native components."
+        exit 1
+    fi
+
+    echo "CoreFX native components successfully built."
+    echo "Product binaries are available at $__BinDir"
+}
+
 __scriptpath=$(cd "$(dirname "$0")"; pwd -P)
+__nativeroot=$__scriptpath/src/Native
 __packageroot=$__scriptpath/packages
 __sourceroot=$__scriptpath/src
 __nugetpath=$__packageroot/NuGet.exe
 __nugetconfig=$__sourceroot/NuGet.Config
+__rootbinpath="$__scriptpath/bin"
 __msbuildpackageid="Microsoft.Build.Mono.Debug"
 __msbuildpackageversion="14.1.0.0-prerelease"
 __msbuildpath=$__packageroot/$__msbuildpackageid.$__msbuildpackageversion/lib/MSBuild.exe
+__BuildArch=x64
+__buildmanaged=false
+__buildnative=false
+# Use uname to determine what the OS is.
+OSName=$(uname -s)
+case $OSName in
+    Linux)
+        __BuildOS=Linux
+        ;;
 
-if [ $(uname) == "Linux" ]; then
-    __monoroot=/usr
-elif [ $(uname) == "FreeBSD" ]; then
-    __monoroot=/usr/local
-else
-    __monoroot=/Library/Frameworks/Mono.framework/Versions/Current
-fi
+    Darwin)
+        __BuildOS=OSX
+        ;;
+
+    FreeBSD)
+        __BuildOS=FreeBSD
+        ;;
+
+    OpenBSD)
+        __BuildOS=OpenBSD
+        ;;
+
+    NetBSD)
+        __BuildOS=NetBSD
+        ;;
+
+    *)
+        echo "Unsupported OS $OSName detected, configuring as if for Linux"
```

Comment:
I know this is how it was, but we should make this an error and remove OpenBSD and NetBSD until there's an actual effort to port to those.